### PR TITLE
sql: add autocommit_before_ddl setting

### LINKED ISSUE #87236: sql: provide a mode to autocommit an open transaction on DDLs
**Is your feature request related to a problem? Please describe.**
Cockroach supports online schema changes, which are cool, but are not perfectly compatible with the transactional schema changes offered by postgres (TODO: associate some issues). Not only are they not compatible in that some side-effects, like adding a new column, do not become visible until after the transaction, but also there are limitations around what can be done in a transaction (https://github.com/cockroachdb/cockroach/issues/49351).

**Describe the solution you'd like**
If you're willing to give up on actually being able to roll back, then just not using a transaction is likely to provide the semantics folks want in many cases. In some cases, we've suggested as much to folks and they've changed their applications. Sometimes changing an application is hard.

It turns out that MySQL, a database which does not support transactional DDL, just implicitly commits any open transaction upon the execution of such a statement (see https://dev.mysql.com/doc/refman/5.7/en/implicit-commit.html).

This issue proposes: 
1) Add a session variable such that DDLs lead to impliclty committing any open transaction
2) Allow COMMIT to executed outside of an open transaction without an error (#87233)

**Describe alternatives you've considered**

We should support transactional schema changes, and we will, but it'll take a while to get there.

**Additional context**

This should help unblock https://github.com/grafana/grafana/issues/8900.

Jira issue: CRDB-19226

Epic CRDB-34097

### PR DESCRIPTION
**Pull request by rafiss**
fixes https://github.com/cockroachdb/cockroach/issues/87236
informs https://github.com/cockroachdb/cockroach/issues/87233

This setting can be used to improve compatibility with some tools that
do not work well due to our limitations with schema changes in explicit
transactions. It also can be used to use schema changes under READ
COMMITTED more easily, without needing to teach the schema changer about
READ COMMITTED.

Release note (sql change): Added the autocommit_before_ddl session
variable. When set to true, any schema change statement that is sent
during an explicit transaction will cause the transaction to commit
before executing the schema change.

---

### sql: handle COMMIT outside of txn when autocommit_before_ddl is on

With this setting on, COMMIT, ROLLBACK, and other statements that
normally cause errors when used outside of an explicit transaction are
instead treated as warnings. This is because the setting can cause a
transaction to end earlier than an application may expect.

### sql: add hint to use autocommit_before_ddl setting

Update the error message returned when using schema changes under READ
COMMITTED to hint towards using the autocommit_before_ddl setting.

### CODE REVIEW

File: pkg/sql/conn_executor_exec.go [hunk 1]
```diff
@@ -2241,11 +2257,13 @@ func (ex *connExecutor) handleTxnRowsWrittenReadLimits(ctx context.Context) erro
 // transaction contains multiple statements, and an upgrade was attempted, an
 // error is returned.
 func (ex *connExecutor) maybeUpgradeToSerializable(ctx context.Context, stmt Statement) error {
+	var txnSchemaChangeErr = pgerror.Newf(
```
Comment (ajwerner) [inline pkg/sql/conn_executor_exec.go:2260]: This should be defined globally as opposed to in this function (which will allocate a new one every time)

**Review by ajwerner — COMMENTED** [review]
It'd be cool to see if grafana now works with cockroach if you set this.

**Review by fqazi — APPROVED** [review]
:lgtm:

> Reviewed 8 of 9 files at r1, 4 of 4 files at r4, 2 of 2 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118440#-:-NpSP3qhD7EBf-6fGPkb:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @rafiss)

___
*[`-- commits` line 21 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/118440#-NpSOA1l31imEOXJvfwo):*
*application
___
*[`pkg/sql/conn_executor_exec.go` line 2258 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/118440#-NpSOi1E7-X1zUlbqF8G:-NpSOi1E7-X1zUlbqF8H:bdrgdjt) ([raw file](https://github.com/cockroachdb/cockroach/blob/58011c708ff1a72fe6b479dc792d4d21c8234979/pkg/sql/conn_executor_exec.go#L2258)):*
> ```Go
> var txnSchemaChangeErr = pgerror.Newf(
> 	pgcode.FeatureNotSupported,
> 	"to use multi-statement transactions involving a schema change under weak isolation levels, enable the autocommit_before_ddl setting",
> ```

Is the plan to make this the default behaviour for read committed out of the box? i.e. Serializable  default to the old behaviour and read committed has this implicitly set

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118440)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi and @michae2)

___
*[`-- commits` line 21 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/118440#-NpSOA1l31imEOXJvfwo):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

*application
</blockquote></details>

fixed
___
*[`pkg/sql/conn_executor_exec.go` line 2258 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/118440#-NpSOi1E7-X1zUlbqF8G:-NpUwwn79GPp0SPOdQpw:b-z1mn9d) ([raw file](https://github.com/cockroachdb/cockroach/blob/58011c708ff1a72fe6b479dc792d4d21c8234979/pkg/sql/conn_executor_exec.go#L2258)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Is the plan to make this the default behaviour for read committed out of the box? i.e. Serializable  default to the old behaviour and read committed has this implicitly set
</blockquote></details>

in the very long term maybe, but there's no plan to do that. for now, we decided to just guide people towards the setting so it stays opt-in, since it's a pretty significant behavior change that people should know what they are signing up for.

**Comment by michae2** [comment]
Was there any particular reason we didn't go with [Ben's suggestion](https://github.com/cockroachdb/cockroach/issues/87236#issuecomment-1234513706) to do COMMIT; DDL; BEGIN; instead of COMMIT; DDL;?

**Pull request merged**